source/util/move_to_front.h

```diff
@@ -0,0 +1,325 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIBSPIRV_UTIL_MOVE_TO_FRONT_H_
+#define LIBSPIRV_UTIL_MOVE_TO_FRONT_H_
+
+#include <cassert>
+#include <cstdint>
+#include <ostream>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+namespace spvutils {
+
+// Log(n) move-to-front implementation. Implements two main functions:
+// IdFromRank - access id by its 1-indexed rank in the move-to-front sequence.
+// RankFromId - get the rank of the given id in the move-to-front sequence.
+// Accessing an id with any of the two functions moves the id to the front
```

Comment:
.... But in this implementation ids are not unrestricted.  You *require* a new Id to be next_id_.

That still seems a weird choice.  How do I know what the next id is?

The current setup seems strangely coupled to its client, and not very reusable.  I'd rather have arbitrary Ids allocated by the client, and only one way to inject a new Id into the MTF sequence.  I'd be happy with an explicit insertion.